src/scipp/operations.py

```diff
@@ -0,0 +1,80 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2022 Scipp contributors (https://github.com/scipp)
+# @author Simon Heybrock
+from typing import Callable, Optional
+from inspect import signature
+from .core import Variable
+from ._scipp.core import experimental_transform
+
+
+def _as_numba_cfunc(function, unit_func=None):
+    import numba
+    dtype = 'double'
+    narg = len(signature(function).parameters)
+    cfunc = numba.cfunc(dtype + '(' + ','.join([dtype] * narg) + ')')(function)
+    cfunc.unit_func = function if unit_func is None else unit_func
+    return cfunc
+
+
+def elementwise_transform(func: Callable,
+                          unit_func: Optional[Callable] = None,
+                          dtype: str = 'float64',
+                          auto_convert_dtypes: bool = False) -> Callable:
+    """
+    Create a function for transforming input variables based on element-wise operation.
+
+    This uses ``numba.cfunc`` to compile a kernel that Scipp can use for transforming
+    the variable contents. Only variables with dtype=float64 are supported. Variances
+    are not supported.
+
+    Custom kernels can reduce intermediate memory consumption and improve performance
+    in multi-step operations with large input variables.
+
+    Parameters
+    ----------
+    func:
+        Function to compute an output element from input element values.
+    unit_func:
+        Function to compute the output unit. If ``None``, ``func`` wil be used.
+    dtype:
+        Must be 'float64'.
```

Comment:
Would this turn into a list if we support heterogeneous dtypes?

Can we use type annotations on `func` instead?